### PR TITLE
rbd: log stdError for cryptosetup commands

### PR DESCRIPTION
**Pull request by Madhu-1**
If we hit any error while running the `cryptosetup` commands we are logging only the error message.  with only an error message it is difficult to analyze the problem, logging the stdError will help us to check what is the problem.

updates: #2610

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

### CODE REVIEW

File: internal/util/crypto.go [hunk 1]
```diff
@@ -249,13 +253,13 @@ func DeviceEncryptionStatus(ctx context.Context, devicePath string) (mappedDevic
 		return devicePath, "", nil
 	}
 	mapPath := strings.TrimPrefix(devicePath, mapperFilePathPrefix+"/")
-	stdout, _, err := LuksStatus(mapPath)
-	if err != nil {
-		log.DebugLog(ctx, "device %s is not an active LUKS device: %v", devicePath, err)
+	stdout, stdErr, err := LuksStatus(mapPath)
+	if err != nil || stdErr != "" {
+		log.DebugLog(ctx, "device %s is not an active LUKS device: %v stdError:%s", devicePath, err, stdErr)
```
Comment (pkalever) [inline internal/util/crypto.go]: Just to keep it consistent just like other, use `: error:%v stdError:%s`

**Review by pkalever — COMMENTED** [review]
minor nit, rest looks good to me

**Review by pkalever — DISMISSED** [review]
It is not mandatory, but could be a great time to substitute `%s` use with `%q`, but can be done later as well :-)

Anyway, the intended changes help us debug the original issue further and this changes LGTM.

Thanks @Madhu-1 :+1:

**Comment by Madhu-1** [comment]
> It is not mandatory, but could be a great time to substitute `%s` use with `%q`, but can be done later as well :-)
> 
Yes, we can use `%q`   for devicePath. sounds good for a follow-up as it needs changes in many places.

> Anyway, the intended changes help us debug the original issue further and this changes LGTM.
> 
> Thanks @Madhu-1 +1

**Comment by pkalever** [comment]
```
I1108 09:13:29.692877   61132 mount_linux.go:463] Attempting to determine if disk "/dev/nbd0" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/nbd0])
I1108 09:13:29.698725   61132 mount_linux.go:466] Output: ""
I1108 09:13:29.698775   61132 crypto.go:199] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad Encrypting device /dev/nbd0 with LUKS
E1108 09:13:31.143934   61132 crypto.go:202] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad failed to encrypt device /dev/nbd0 with LUKS: error:an error (exit status 1) occurred while running cryptsetup args: [-q luksFormat --type luks2 --hash sha256 /dev/nbd0 -d /dev/stdin] stdError:Cannot format device /dev/nbd0.
E1108 09:13:31.143987   61132 encryption.go:204] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad failed to encrypt volume replicapool/csi-vol-1f2393db-4074-11ec-b005-5a8876ad69ad: an error (exit status 1) occurred while running cryptsetup args: [-q luksFormat --type luks2 --hash sha256 /dev/nbd0 -d /dev/stdin]
I1108 09:13:31.692169   61132 cephcmds.go:62] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad command succeeded: rbd [unmap /dev/nbd0 --device-type nbd]
E1108 09:13:31.692457   61132 utils.go:186] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad GRPC error: rpc error: code = Internal desc = failed to encrypt rbd image replicapool/csi-vol-1f2393db-4074-11ec-b005-5a8876ad69ad: failed to encrypt volume replicapool/csi-vol-1f2393db-4074-11ec-b005-5a8876ad69ad: an error (exit status 1) occurred while running cryptsetup args: [-q luksFormat --type luks2 --hash sha256 /dev/nbd0 -d /dev/stdin]
I1108 09:13:32.203278   61132 utils.go:177] ID: 10 GRPC call: /csi.v1.Node/NodeGetCapabilities
```


E1108 09:13:31.143934   61132 crypto.go:202] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad failed to encrypt device /dev/nbd0 with LUKS: error:an error (exit status 1) occurred while running cryptsetup args: [-q luksFormat --type luks2 --hash sha256 /dev/nbd0 -d /dev/stdin] stdError:Cannot format device /dev/nbd0.

@Madhu-1 @nixpanic @humblec 
Is this anything related to https://unix.stackexchange.com/questions/232123/cannot-format-device-dev-sda4-which-is-still-in-use ?

I don't think there is any chance for having a stagedTargetPath created, but adding here for more eyes

**Comment by Madhu-1** [comment]
> ```
> I1108 09:13:29.692877   61132 mount_linux.go:463] Attempting to determine if disk "/dev/nbd0" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/nbd0])
> I1108 09:13:29.698725   61132 mount_linux.go:466] Output: ""
> I1108 09:13:29.698775   61132 crypto.go:199] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad Encrypting device /dev/nbd0 with LUKS
> E1108 09:13:31.143934   61132 crypto.go:202] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad failed to encrypt device /dev/nbd0 with LUKS: error:an error (exit status 1) occurred while running cryptsetup args: [-q luksFormat --type luks2 --hash sha256 /dev/nbd0 -d /dev/stdin] stdError:Cannot format device /dev/nbd0.
> E1108 09:13:31.143987   61132 encryption.go:204] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad failed to encrypt volume replicapool/csi-vol-1f2393db-4074-11ec-b005-5a8876ad69ad: an error (exit status 1) occurred while running cryptsetup args: [-q luksFormat --type luks2 --hash sha256 /dev/nbd0 -d /dev/stdin]
> I1108 09:13:31.692169   61132 cephcmds.go:62] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad command succeeded: rbd [unmap /dev/nbd0 --device-type nbd]
> E1108 09:13:31.692457   61132 utils.go:186] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad GRPC error: rpc error: code = Internal desc = failed to encrypt rbd image replicapool/csi-vol-1f2393db-4074-11ec-b005-5a8876ad69ad: failed to encrypt volume replicapool/csi-vol-1f2393db-4074-11ec-b005-5a8876ad69ad: an error (exit status 1) occurred while running cryptsetup args: [-q luksFormat --type luks2 --hash sha256 /dev/nbd0 -d /dev/stdin]
> I1108 09:13:32.203278   61132 utils.go:177] ID: 10 GRPC call: /csi.v1.Node/NodeGetCapabilities
> ```
> 
> E1108 09:13:31.143934 61132 crypto.go:202] ID: 9 Req-ID: 0001-0024-7d08c2d2-55cd-4626-b14d-074f633f26b1-0000000000000002-1f2393db-4074-11ec-b005-5a8876ad69ad failed to encrypt device /dev/nbd0 with LUKS: error:an error (exit status 1) occurred while running cryptsetup args: [-q luksFormat --type luks2 --hash sha256 /dev/nbd0 -d /dev/stdin] stdError:Cannot format device /dev/nbd0.
> 
> @Madhu-1 @nixpanic @humblec Is this anything related to https://unix.stackexchange.com/questions/232123/cannot-format-device-dev-sda4-which-is-still-in-use ?
> 
> I don't think there is any chance for having a stagedTargetPath created, but adding here for more eyes

from the logs looks like this is the first request for this volume and also it does not log in-use error message? not sure what is the exact reason

**Comment by Madhu-1** [comment]
```
[1mSTEP[0m: perform IO on rbd-nbd volume after nodeplugin restart
E1108 09:19:50.840885   83678 util.go:247] kernel 4.19.202
 does not support required features
Nov  8 09:19:50.841: INFO: waiting for kubectl (delete -f args []) to finish
Nov  8 09:19:50.841: INFO: Running '/usr/bin/kubectl --server=https://192.168.39.234:8443 --kubeconfig=/root/.kube/config --namespace=cephcsi-e2e-f1c20812 delete -f -'
Nov  8 09:19:50.975: INFO: stderr: "warning: deleting cluster-scoped resources, not scoped to the provided namespace\n"
Nov  8 09:19:50.975: INFO: stdout: "storageclass.storage.k8s.io \"csi-rbd-sc\" deleted\n"
Nov  8 09:19:50.980: INFO: ExecWithOptions {Command:[/bin/sh -c ceph fsid] Namespace:rook-ceph PodName:rook-ceph-tools-7467d8bf8-jmbsg ContainerName:rook-ceph-tools Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Nov  8 09:19:50.980: INFO: >>> kubeConfig: /root/.kube/config
Nov  8 09:19:53.306: INFO: Waiting up to &PersistentVolumeClaim{ObjectMeta:{rbd-pvc  rbd-694    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*csi-rbd-sc,VolumeMode:nil,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},},} to be in Bound state
Nov  8 09:19:53.306: INFO: waiting for PVC rbd-pvc (0 seconds elapsed)
Nov  8 09:19:55.310: INFO: waiting for PVC rbd-pvc (2 seconds elapsed)
Nov  8 09:19:55.317: INFO: Waiting for PV pvc-4d94f07a-aea1-46a7-927a-4cf2e8155537 to bind to PVC rbd-pvc
Nov  8 09:19:55.317: INFO: Waiting up to timeout=10m0s for PersistentVolumeClaims [rbd-pvc] to have phase Bound
Nov  8 09:19:55.319: INFO: PersistentVolumeClaim rbd-pvc found and phase=Bound (2.516305ms)
Nov  8 09:19:55.319: INFO: Waiting up to 10m0s for PersistentVolume pvc-4d94f07a-aea1-46a7-927a-4cf2e8155537 to have phase Bound
Nov  8 09:19:55.322: INFO: PersistentVolume pvc-4d94f07a-aea1-46a7-927a-4cf2e8155537 found and phase=Bound (2.824542ms)
Nov  8 09:19:55.334: INFO: Waiting up to csi-rbd-demo-pod to be in Running state
Nov  8 09:20:07.344: INFO: ExecWithOptions {Command:[/bin/sh -c sync /var/lib/www/html] Namespace:rbd-694 PodName:csi-rbd-demo-pod ContainerName:web-server Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Nov  8 09:20:07.344: INFO: >>> kubeConfig: /root/.kube/config
Nov  8 09:20:07.498: INFO: stdErr occurred: sync: error syncing '/var/lib/www/html': Input/output error

Nov  8 09:20:07.498: FAIL: failed to sync, err: command terminated with exit code 1, stdErr: sync: error syncing '/var/lib/www/html': Input/output error
```

@pkalever i also see input/output error for plugin restart test case.

**Comment by pkalever** [comment]
> @pkalever i also see input/output error for plugin restart test case.

@Madhu-1 
It is already tracked with https://github.com/ceph/ceph-csi/issues/2204

File: internal/util/crypto.go [hunk 2]
```diff
@@ -197,19 +197,20 @@ func VolumeMapper(volumeID string) (mapperFile, mapperFilePath string) {
 // EncryptVolume encrypts provided device with LUKS.
 func EncryptVolume(ctx context.Context, devicePath, passphrase string) error {
 	log.DebugLog(ctx, "Encrypting device %s with LUKS", devicePath)
-	if _, _, err := LuksFormat(devicePath, passphrase); err != nil {
-		return fmt.Errorf("failed to encrypt device %s with LUKS: %w", devicePath, err)
+	_, stdErr, err := LuksFormat(devicePath, passphrase)
+	if err != nil || stdErr != "" {
+		log.ErrorLog(ctx, "failed to encrypt device %s with LUKS: error:%v stdError:%s", devicePath, err, stdErr)
```
Comment (nixpanic) [inline internal/util/crypto.go]: nit: the logging gets quite ugly this way, maybe do

    failed to encrypt device %q with LUKS (%v): %s

Comment (Madhu-1) [reply internal/util/crypto.go]: Done. logging with same format.

File: internal/util/crypto.go [hunk 3]
```diff
@@ -196,40 +196,44 @@ func VolumeMapper(volumeID string) (mapperFile, mapperFilePath string) {
...
 // OpenEncryptedVolume opens volume so that it can be used by the client.
 func OpenEncryptedVolume(ctx context.Context, devicePath, mapperFile, passphrase string) error {
-	log.DebugLog(ctx, "Opening device %s with LUKS on %s", devicePath, mapperFile)
-	_, stderr, err := LuksOpen(devicePath, mapperFile, passphrase)
-	if err != nil {
-		log.ErrorLog(ctx, "failed to open LUKS device %q: %s", devicePath, stderr)
+	log.DebugLog(ctx, "Opening device %q with LUKS on %q", devicePath, mapperFile)
+	_, stdErr, err := LuksOpen(devicePath, mapperFile, passphrase)
+	if err != nil || stdErr != "" {
+		log.ErrorLog(ctx, "failed to open device %q: (%v): %s", devicePath, err, stdErr)
```
Comment (nixpanic) [inline internal/util/crypto.go]: nit: unneeded `:` before `(%v)`

**Review by nixpanic — DISMISSED** [review]

**Review by pkalever — DISMISSED** [review]
Looks clean :-)

File: internal/util/crypto.go [hunk 4]
```diff
@@ -196,40 +196,44 @@ func VolumeMapper(volumeID string) (mapperFile, mapperFilePath string) {
 
 // EncryptVolume encrypts provided device with LUKS.
 func EncryptVolume(ctx context.Context, devicePath, passphrase string) error {
-	log.DebugLog(ctx, "Encrypting device %s with LUKS", devicePath)
-	if _, _, err := LuksFormat(devicePath, passphrase); err != nil {
-		return fmt.Errorf("failed to encrypt device %s with LUKS: %w", devicePath, err)
+	log.DebugLog(ctx, "Encrypting device %q	 with LUKS", devicePath)
+	_, stdErr, err := LuksFormat(devicePath, passphrase)
+	if err != nil || stdErr != "" {
+		log.ErrorLog(ctx, "failed to encrypt device %q with LUKS (%v) %s", devicePath, err, stdErr)
```
Comment (nixpanic) [inline internal/util/crypto.go]: lol, almost the same nit as before: I'd prefer to have `(%v): %s` with the `:` after `(%v)`... You had it earlier, but dropped it  as well when I asked to drop the `:` _before_ `(%v)`.

**Review by nixpanic — DISMISSED** [review]

File: internal/util/crypto.go (hunk 4 above)
Comment (Madhu-1) [reply internal/util/crypto.go]: :D pushed new changes. need approval again

**Review by nixpanic — APPROVED** [review]

**Review by pkalever — APPROVED** [review]

**Comment by Madhu-1** [comment]
@Mergifyio rebase

**Comment by Rakshith-R** [comment]
@mergifyio rebase

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by pkalever** [comment]
/retest ci/centos/mini-e2e-helm/k8s-1.20


```
X Exiting due to GUEST_PROVISION: Failed to start host: creating host: create: Error creating machine: Error in driver during machine creation: IP not available after waiting: machine minikube didn't return IP after 1 minute

```

**Comment by nixpanic** [comment]
@mergifyio rebase

**Pull request merged**